tests/agent_server/test.py

```diff
@@ -0,0 +1,293 @@
+            {
+                "key": "key2",
+                "value": f"value{i}",
+                "id": res_id_2,
+                "send_event": True,
+                "requires": [res_id_3],
+                "purged": False,
+            },
+            {
+                "key": "key3",
+                "value": "value",
+                "id": res_id_3,
+                "send_event": True,
+                "requires": [],
+                "purged": False,
+            },
+        ]
+
+        result = await client.put_version(
+            tid=environment,
+            version=version,
+            resources=resources,
+            unknowns=[],
+            version_info={},
+            compiler_version=get_compiler_version(),
+        )
+        assert result.code == 200
+
+        # do a deploy
+        result = await client.release_version(environment, version, True, const.AgentTriggerMethod.push_full_deploy)
+        assert result.code == 200
+
+        result = await client.get_version(environment, version)
+        assert result.code == 200
+
+        await _wait_until_deployment_finishes(client, environment, version)
+
+    async def get_last_change(
+        resource_type: str,
+        agent: str,
+        key: str,
+        value: str,
+        before: Optional[datetime.datetime] = None,
+    ):
+        before = before or datetime.datetime.now()
+
+        while before is not None:
+            result = await client.get_resource_actions(
+                tid=environment,
+                resource_type=resource_type,
+                agent=agent,
+                attribute=key,
+                attribute_value=value,
+                limit=1,
+                last_timestamp=before,
+            )
+            assert result.code == 200
+
+            actions = result.result.get("data", [])
+
+            for action in actions:
+                if action.get("action", "") != "deploy":
+                    continue
+                change = action.get("change") or "nochange"
+                if change != "nochange":
+                    return {
+                        "change": change,
+                        "finished": action.get("finished"),
+                        "version": action.get("version"),
+                    }
+
+            if len(actions) == 0:
+                before = None
+            else:
+                before = actions[-1].get("started")
+
+        return None
+
+    async def get_first_change(
+        resource_type: str,
+        agent: str,
+        key: str,
+        value: str,
+        after: Optional[datetime.datetime] = None,
+    ):
+        after = after or datetime.datetime.fromtimestamp(0)
+
+        while after is not None:
+            result = await client.get_resource_actions(
+                tid=environment,
+                resource_type=resource_type,
+                agent=agent,
+                attribute=key,
+                attribute_value=value,
+                limit=1,
```

Comment:
getting them one-by-one is very inefficient.

Reply:
Obviously yes, this was only for the prototype, to demonstrate that it works, in the real MR, I picked 25 as size, we could maybe make it even bigger

Reply:
I would also test it quite extensively here, because testing in this context is a lot easier. 
So I would make it mostly identical to the eventual version

Reply:
I propose 3: https://github.com/inmanta/inmanta-core/pull/2762/files#diff-95320f17c120fb5f0de34426ebfaa0e937201dc0b9a90a293ecd0667ccaaaf2fR39